### PR TITLE
Fix source map behaviour in Parcel and move node_modules compiled extraction to new parcel transformer

### PR DESCRIPTION
**Pull request by JakeLane**
The current behaviour for the compiled parcel transformer does not handle source maps correctly. This PR resolves these issues by using the source maps correctly and introducing a new parcel transformer to handle external compiled deps correctly.

### CODE REVIEW

**Review by marcins — COMMENTED** [review]
I've gone through this PR and conceptually it seems fine to me - there's a fair bit of change, and I don't have a huge amount of context, so it wasn't necessarily easily to mentally parse the changes.

The only comment I had when looking through was whether there are is any opportunity to pass existing sourcemaps into some of the Babel methods (`generate` or `transform`) that might save some of the manual merging / manipulation after, but you likely already considered that and there were reasons it wouldn't work.

File: packages/parcel-transformer-external/src/__tests__/transformer.parceltest.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+const workerFarm = createWorkerFarm();
+
+afterAll(() => {
+  workerFarm.end();
+});
+
+const outputFS = new MemoryFS(workerFarm);
+
+const getParcelInstance = (workingDir: string, sourceMap = false) => {
```
Comment (dddlr) [inline packages/parcel-transformer-external/src/__tests__/transformer.parceltest.ts:17]: i don't have enough context on the Parcel stuff to properly review it beyond "seems fine to me", but i'll do what i can to review the parts that are Compiled related

File: packages/parcel-transformer/src/__tests__/transformer.parceltest.ts [hunk 2]
```diff
@@ -15,15 +18,35 @@ afterAll(() => {
...
 
-const getParcelInstance = (workingDir: string) => {
+function findTargetSourcePosition(source: string, regex: RegExp): { line: number; column: number } {
+  const lines = source.split('\n');
+
+  for (let i = 0; i < lines.length; i++) {
+    const line = lines[i];
+    const match = line.match(regex);
+
+    if (match && match.index != null) {
```
Comment (dddlr) [inline packages/parcel-transformer/src/__tests__/transformer.parceltest.ts:30]: is there a reason loose equality is used here and not strict equality?

**Comment by dddlr** [comment]
>  introducing a new parcel transformer to handle external compiled deps correctly

i'm still reading through and parsing through the PR 😅 but what was the reason behind taking some of the functionality of `@compiled/parcel-transformer` and putting them into a new package `@compiled/parcel-transformer-external`? is it to avoid bogging down `@compiled/parcel-transformer` with too much complexity?

File: package.json [hunk 3]
```diff
@@ -45,7 +45,7 @@
     "test": "jest --no-cache",
     "test:cover": "yarn test --collectCoverage",
     "test:imports": "node test/test-imports",
-    "test:parcel": "jest --testMatch '**/src/**/*.parceltest.{ts,tsx}'",
+    "test:parcel": "jest --testMatch '**/src/**/*.parceltest.{ts,tsx}' --testEnvironment=node",
```
Comment (dddlr) [inline package.json:48]: does it not work without this change?

File: packages/parcel-transformer/src/index.ts [hunk 4]
```diff
@@ -148,13 +129,16 @@ export default new Transformer<ParcelTransformerOpts>({
     const includedFiles: string[] = [];
     const code = asset.isASTDirty() ? undefined : await asset.getCode();
 
+    const shouldUseSourceMaps = asset.env.sourceMap != null;
```
Comment (dddlr) [inline packages/parcel-transformer/src/index.ts:132]: wouldn't `!!shouldUseSourceMaps` be clearer?

**Comment by JakeLane** [comment]
> I've gone through this PR and conceptually it seems fine to me - there's a fair bit of change, and I don't have a huge amount of context, so it wasn't necessarily easily to mentally parse the changes.
> 
> The only comment I had when looking through was whether there are is any opportunity to pass existing sourcemaps into some of the Babel methods (`generate` or `transform`) that might save some of the manual merging / manipulation after, but you likely already considered that and there were reasons it wouldn't work.

The main goal for the new transformer is to avoid running the compiled babel plugin for external node modules when not required. Bringing in babel context would be a concern for that goal.

I think we can reuse the AST in other babel contexts. I think that's a perf win on the table but I didn't want to make this change more complicated than it already is. I'll create a ticket on the backlog to action when we can.

File: packages/parcel-transformer/src/__tests__/transformer.parceltest.ts (hunk 2 above)
Comment (JakeLane) [reply packages/parcel-transformer/src/__tests__/transformer.parceltest.ts:30]: These are borrowed from a different implementation. I'll refactor to remove the sketchy null comparison

File: package.json (hunk 3 above)
Comment (JakeLane) [reply package.json:48]: yep this is required

**Review by dddlr — APPROVED** [review]
looks good!

**Review by dddlr — APPROVED** [review]

**Review by liamqma — APPROVED** [review]

**Pull request merged**